ethexe/contracts/src/Mirror.sol

```diff
@@ -172,14 +176,105 @@ contract Mirror is IMirror {
+                    bytes32 topic1;
+                    assembly ("memory-safe") {
+                        topic1 := calldataload(add(payload.offset, 1))
+                    }
+
+                    if (
+                        topic1 != StateChanged.selector && topic1 != MessageQueueingRequested.selector
+                            && topic1 != ReplyQueueingRequested.selector && topic1 != ValueClaimingRequested.selector
+                            && topic1 != ExecutableBalanceTopUpRequested.selector && topic1 != Message.selector
+                            && topic1 != Reply.selector && topic1 != ValueClaimed.selector
+                    ) {
+                        uint256 size;
+                        unchecked {
+                            size = payload.length - topicsLengthInBytes;
+                        }
+
+                        uint256 memPtr = Memory.allocate(size);
+                        assembly ("memory-safe") {
+                            calldatacopy(memPtr, add(payload.offset, topicsLengthInBytes), size)
+                        }
+
+                        bytes32 topic2;
+                        bytes32 topic3;
+                        bytes32 topic4;
+                        assembly ("memory-safe") {
+                            topic2 := calldataload(add(payload.offset, 33))
+                            topic3 := calldataload(add(payload.offset, 65))
+                            topic4 := calldataload(add(payload.offset, 97))
+                        }
+
+                        if (topicsLength == 1) {
+                            assembly ("memory-safe") {
+                                log1(memPtr, size, topic1)
+                            }
+                        } else if (topicsLength == 2) {
+                            assembly ("memory-safe") {
+                                log2(memPtr, size, topic1, topic2)
+                            }
+                        } else if (topicsLength == 3) {
+                            assembly ("memory-safe") {
+                                log3(memPtr, size, topic1, topic2, topic3)
+                            }
+                        } else if (topicsLength == 4) {
+                            assembly ("memory-safe") {
+                                log4(memPtr, size, topic1, topic2, topic3, topic4)
+                            }
+                        }
+
+                        return;
+                    }
+                }
+            }
+        }
+
         emit Message(_message.id, _message.destination, _message.payload, _message.value);
     }
 
     /// @dev Non-zero value always sent since never goes to mailbox.
     function _sendReplyMessage(Gear.Message calldata _message) private {
         _transferValue(_message.destination, _message.value);
 
-        emit Reply(_message.payload, _message.value, _message.replyDetails.to, _message.replyDetails.code);
+        if (_message.destination.code.length > 0) {
```

Comment:
it may no longer work in next hardfork

Reply:
> why is it?

https://github.com/ethereum/EIPs/blob/master/EIPS/eip-7702.md
see `extcodesize` changes. it's no longer possible to know is address user or contract.
`extcodesize` has also been removed in `EOFv1` smart contracts.

> shouldn't this check be == 4?

don't get it

Reply:
oh, for some reason on the first sight though that its about reply code ...